File: greedy algorithm/imbalance/imbalance.py
def new_pairs(species, amnt_chambers, pairs, j):
    for i in range(amnt_chambers):
        pairs[j] = species[i]
        pairs[j+1] = species[amnt_chambers*2 - 1 - i]
        j+=2

def calc_imbalance(amnt_chambers, pairs, AM):
    imbalance, result, CM = 0, 0, 0

    for i in range(0, amnt_chambers*2, 2):
        CM = pairs[i] + pairs[i+1]
        result = abs(CM - AM)
        imbalance += result

    return imbalance

if __name__ == "__main__":
    amnt_chambers, amnt_species, sets = 0, 0, 0
    while True:
        try:    
            entry = list(map(int,input().split()))
            amnt_chambers = entry[0]
            amnt_species = entry[1]
            species = []
            sets += 1
            AM = 0

            new_species = list(map(int,input().split()))
            for i in new_species: 
                species.append(i)
                AM += i
            
            AM = AM / amnt_chambers

            aux = amnt_species

            while(aux < amnt_chambers*2):
                species.append(0)
                aux += 1

            species.sort()

            pairs = [0]*(amnt_chambers*2)

            new_pairs(species, amnt_chambers, pairs, 0)

            print("Set #", sets, sep='')
            for i in range(0, amnt_chambers*2, 2):
                print(" ", int(i/2), ": ",sep ='', end="")
                if pairs[i]>0: print(pairs[i], end=" ")
                if pairs[i+1]>0: print(pairs[i+1], end="")
                print()

            imbalance = calc_imbalance(amnt_chambers, pairs, AM)
            print(f'IMBALANCE = {imbalance:,.5f}', '\n')

        except EOFError:
            break

File: greedy algorithm/imbalance/imbalance.cpp
//Student: Gustavo Sampaio Lima
//ID: 12623992

#include <bits/stdc++.h>
#include <iostream>
using namespace std;

void new_pairs(vector<int> species, int amnt_chambers, vector<int> &pairs);
float calc_imbalance(int amnt_chambers, vector<int> pairs, float AM);

int main(){
    int amnt_chambers = 0, amnt_species = 0, new_specie = 0, sets = 0;
    
    while(cin >> amnt_chambers >> amnt_species){
        vector<int> species;
        sets++;

        float AM = 0;

        for(int i = 0; i < amnt_species; i++){
            cin >> new_specie;
            species.push_back(new_specie);
            AM += new_specie;
        }

        AM = AM / amnt_chambers;
        
        int aux = amnt_species;
        while(aux++ < amnt_chambers*2)
            species.push_back(0);
            
        sort(species.begin(), species.end());
        
        vector<int> pairs(amnt_chambers*2);
        new_pairs(species, amnt_chambers, pairs);

        cout << "Set #" << sets << endl;
        for(int i = 0; i < amnt_chambers*2; i+=2){
            cout << " " << i/2 << ": ";
            if(pairs[i] > 0) cout << pairs[i] << " " ;
            if(pairs[i+1] > 0) cout << pairs[i+1];
            cout << endl;
        }

        cout << "IMBALANCE = ";
        float imbalance = calc_imbalance(amnt_chambers, pairs, AM);
        cout.precision(5);
        cout << fixed << imbalance << endl << endl; 
    }

    return 0;
}

void new_pairs(vector<int> species, int amnt_chambers, vector<int> &pairs){
    int j = 0;
    for(int i = 0; i < amnt_chambers; i++){
        pairs[j] = species[i];
        pairs[j+1] = species[amnt_chambers*2 - 1 - i];
        j += 2;
    }
}

float calc_imbalance(int amnt_chambers, vector<int> pairs, float AM){
    float imbalance = 0, result = 0, CM = 0;    

    for(int i = 0; i < amnt_chambers*2; i += 2){
        CM = pairs[i] + pairs[i + 1];
        result = CM - AM;
        if(result < 0) result *= (-1);
        imbalance += result;
    }

    return imbalance;
}